tket/tests/test_LexiRoute.cpp

```diff
@@ -454,10 +461,244 @@ SCENARIO("Test LexiRoute::solve") {
   }
 }
 
+SCENARIO("Test LabellingRoutingMethod") {
+  std::vector<Node> nodes = {
+      Node("test_node", 0), Node("test_node", 1), Node("test_node", 2),
+      Node("node_test", 3), Node("node_test", 4)};
+
+  // straight line
+  Architecture architecture(
+      {{nodes[0], nodes[1]},
+       {nodes[1], nodes[2]},
+       {nodes[2], nodes[3]},
+       {nodes[3], nodes[4]}});
+  ArchitecturePtr shared_arc = std::make_shared<Architecture>(architecture);
+  GIVEN("No qubit to label, empty frontier, check_method.") {
+    Circuit circ(5);
+    std::shared_ptr<MappingFrontier> mf =
+        std::make_shared<MappingFrontier>(circ);
+    LabellingRoutingMethod lrm;
+    REQUIRE(!lrm.check_method(mf, shared_arc));
+  }
+  GIVEN("No qubit to label, partially filled frontier, check_method.") {
+    Circuit circ(5);
+    std::vector<Qubit> qubits = circ.all_qubits();
+    circ.add_op<UnitID>(OpType::CX, {qubits[0], qubits[4]});
+    circ.add_op<UnitID>(OpType::CZ, {qubits[1], qubits[2]});
+    circ.add_op<UnitID>(OpType::ZZPhase, 0.3, {qubits[3], qubits[0]});
+    std::map<UnitID, UnitID> rename_map = {
+        {qubits[0], nodes[0]},
+        {qubits[1], nodes[1]},
+        {qubits[2], nodes[2]},
+        {qubits[3], nodes[3]},
+        {qubits[4], nodes[4]}};
+    circ.rename_units(rename_map);
+    std::shared_ptr<MappingFrontier> mf =
+        std::make_shared<MappingFrontier>(circ);
+    LabellingRoutingMethod lrm;
+    REQUIRE(!lrm.check_method(mf, shared_arc));
+  }
+  GIVEN("Qubit to label, but casually restricted, check_method.") {
+    Circuit circ(5);
+    std::vector<Qubit> qubits = circ.all_qubits();
+    circ.add_op<UnitID>(OpType::CX, {qubits[0], qubits[4]});
+    circ.add_op<UnitID>(OpType::CZ, {qubits[1], qubits[2]});
+    circ.add_op<UnitID>(OpType::ZZPhase, 0.3, {qubits[3], qubits[0]});
+    std::map<UnitID, UnitID> rename_map = {
+        {qubits[0], nodes[0]},
+        {qubits[1], nodes[1]},
+        {qubits[2], nodes[2]},
+        {qubits[4], nodes[4]}};
+    circ.rename_units(rename_map);
+    std::shared_ptr<MappingFrontier> mf =
+        std::make_shared<MappingFrontier>(circ);
+    LabellingRoutingMethod lrm;
+    REQUIRE(!lrm.check_method(mf, shared_arc));
+  }
+  GIVEN(
+      "Two Qubit to label in future slice, causally restricted, "
+      "check_method.") {
+    Circuit circ(5);
+    std::vector<Qubit> qubits = circ.all_qubits();
+    circ.add_op<UnitID>(OpType::CX, {qubits[0], qubits[1]});
+    circ.add_op<UnitID>(OpType::CZ, {qubits[1], qubits[2]});
+    circ.add_op<UnitID>(OpType::CZ, {qubits[2], qubits[3]});
+    circ.add_op<UnitID>(OpType::ZZPhase, 0.3, {qubits[3], qubits[4]});
+    std::map<UnitID, UnitID> rename_map = {
+        {qubits[0], nodes[0]}, {qubits[1], nodes[1]}, {qubits[2], nodes[2]}};
+    circ.rename_units(rename_map);
+    std::shared_ptr<MappingFrontier> mf =
+        std::make_shared<MappingFrontier>(circ);
+    LabellingRoutingMethod lrm;
+    REQUIRE(!lrm.check_method(mf, shared_arc));
```

Comment:
I am not sure I understand why this is not labelling qubit 3 and 4?

Reply:
qubit[3] and qubit[4] are causally restricted by the qubits beforehand, so the relevant gate is not in the MappingFrontier object. `advance_frontier_boundary` would need to be run to clear enough of the gates for them to be relabelled. This is intended behaviour.